Applications/Spire/Source/Playback/PlaybackControlPanel.cpp

```diff
@@ -0,0 +1,425 @@
+#include "Spire/Playback/PlaybackControlPanel.hpp"
+#include "Spire/Spire/Dimensions.hpp"
+#include "Spire/Spire/FieldValueModel.hpp"
+#include "Spire/Ui/DateBox.hpp"
+#include "Spire/Ui/Icon.hpp"
+#include "Spire/Ui/ToggleButton.hpp"
+#include "Spire/Ui/Tooltip.hpp"
+
+using namespace Beam::TimeService;
+using namespace boost;
+using namespace boost::gregorian;
+using namespace boost::posix_time;
+using namespace boost::signals2;
+using namespace Spire;
+using namespace Spire::Styles;
+
+namespace {
+  const auto TIME_TOLERANCE = milliseconds(200);
+
+  auto make_playback_icon_button(const QString& icon_path,
+      const QString& tooltip) {
+    auto button = make_icon_button(
+      imageFromSvg(icon_path, scale(26, 26)), tooltip);
+    button->setFixedSize(scale(26, 26));
+    return button;
+  }
+
+  auto make_play_toggle_button() {
+    auto play_toggle = make_icon_toggle_button(
+      imageFromSvg(":/Icons/play.svg", scale(26, 26)),
+      imageFromSvg(":/Icons/pause.svg", scale(26, 26)),
+      std::make_shared<LocalBooleanModel>(false));
+    play_toggle->setFixedSize(scale(26, 26));
+    add_tooltip(QObject::tr("Play (Space)"), *play_toggle);
+    update_style(*play_toggle, [] (auto& style) {
+      style.get(Checked() > is_a<Icon>()).set(Fill(QColor(0x535353)));
+      style.get((Checked() && Disabled()) > is_a<Icon>()).
+        set(Fill(QColor(0xC8C8C8)));
+    });
+    return play_toggle;
+  }
+
+  struct PlaybackDateModel : public OptionalDateModel {
+    std::shared_ptr<ValueModel<ptime>> m_source;
+    optional<date> m_min_date;
+    LocalValueModel<optional<date>> m_value;
+    QValidator::State m_state;
+    scoped_connection m_connection;
+
+    PlaybackDateModel(std::shared_ptr<ValueModel<ptime>> source,
+        optional<date> min_date)
+        : m_source(std::move(source)),
+          m_min_date(std::move(min_date)),
+          m_value(m_source->get().date()),
+          m_state(test(m_value.get())) {
+      m_connection = m_source->connect_update_signal(
+        std::bind_front(&PlaybackDateModel::on_update, this));
+    }
+
+    optional<date> get_minimum() const override {
+      return m_min_date;
+    }
+
+    optional<date> get_maximum() const override {
+      return day_clock::local_day();
+    }
+
+    QValidator::State get_state() const override {
+      return m_state;
+    }
+
+    const optional<date>& get() const override {
+      return m_value.get();
+    }
+
+    QValidator::State test(const optional<date>& value) const override {
+      if(!value || get_minimum() && value < get_minimum() ||
+          value > get_maximum()) {
+        return QValidator::Invalid;
+      }
+      return QValidator::Acceptable;
+    }
+
+    QValidator::State set(const optional<date>& value) override {
+      m_state = test(value);
+      if(m_state == QValidator::Invalid) {
```

Comment:
A model can never be in an invalid state. It can only be in an intermediate or valid state.